packages/next/src/server/app-render/app-render.tsx

```diff
@@ -1226,6 +1229,10 @@ async function renderToHTMLOrFlightImpl(
   // It got here, which means it did not reject, so clear the timeout to avoid
   // it from rejecting again (which is a no-op anyways).
   clearTimeout(timeout)
+  const hasCapturedError = digestErrorsMap.size > 0
+  const firstCapturedError = hasCapturedError
+    ? digestErrorsMap.values().next().value
+    : null
```

Comment:
```suggestion
  const buildFailingError = digestErrorsMap.size > 0
    ? digestErrorsMap.values().next().value
    : null
```

Consider just capturing the error and not trying to synchronize the `hasError` and `firstError` bindings. Then everywhere you used to check for the errors size or hasError just check if `buildFailingError` exists